doctr: init at 0.7.0. Also dependencies mplcursors and pypdfium2

nh2 opened the pull request:
* [ ] :arrow_right: Looking for co-maintainers! :arrow_left:

[doctr](https://mindee.github.io/doctr/) is a pretty good OCR software.

## TODO

This is a draft PR because:

* [x] The distro packaging story for the dependency `pypdfium2` isn't fully figured out yet: https://github.com/pypdfium2-team/pypdfium2/issues/272#issuecomment-1793833094
  * The upstream build system currently does various network acceses during builds.
  * My patches provided fix that (via the standard solution of passing in everything that needs to be downloaded), but it would be better if they were smaller or not needed.
  * **Edit:** Solved now, as per new upstream instructions.

Still, already sharing this PR in case others want to use `doctr`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - [x] Tested the `doctr` example code to detect text from a scanned bank account statement
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

nh2 committed:
doctr: init at 0.7.0.

Also includes required dependencies.

mara004 commented:
> The distro packaging story for the dependency pypdfium2 isn't fully figured out yet
> * The upstream build system currently does various network acceses during builds.
> * My patches provided fix that (via the standard solution of passing in everything that needs to be downloaded), but it would be better if they were smaller or not needed.

pypdfium2 TLDR (maintainer POV)
* It should be doable without patches by externalizing the data files generation instead of trying to put it in our upstream setupsrc (see new instructions https://github.com/pypdfium2-team/pypdfium2/#install-source-caller)
* Packaging pdfium separately first may be more advisable for a distro, which somewhat changes the pypdfium2 packaging story anyway (likely results in a different layout than a pdfium-binaries tarball - e.g. where would version info be stored at system level, and in which form?)

---

That said, here's another idea for what you're currently doing (bundling): Maybe you could just use one of our published wheels to source the data files? The prebuilt tarball's origin shouldn't make much difference (I guess?, anyway), but would relieve the necessity of dealing with ctypesgen/bindings and (most of) the version info.

nh2 commented:
I updated the `pypdfium2-bin` build as per upstream instructions: https://github.com/pypdfium2-team/pypdfium2/tree/30c60af438b7cd90e22d42dd2ba5bffdeb568c42#install-source-caller

I also added a test that checks that `doctr` successfully detects text.

This is ready for review now.

mara004 commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,152 @@
+    "${ctypesgen_pypdfium_fork}/bin/ctypesgen" \
+      --library pdfium \
+      --compile-libdirs "${pdfium-bin}/lib" \
+      --runtime-libdirs "${pdfium-bin}/lib" \
```
I think `--runtime-libdirs` should be `.`, supposing `${pdfium-bin}` is not available at runtime?
Or if it is, you wouldn't have to copy in the binary below.

nh2 replied:
@mara004 It seems to work with both `--runtime-libdirs "${pdfium-bin}/lib"` and `--runtime-libdirs .`.

With `--runtime-libdirs "${pdfium-bin}/lib"`, the generated `lib/python3.11/site-packages/pypdfium2_raw/bindings.py` contains:

```python
_libdirs = ['/nix/store/q79y5h4af3vyi5kqbrwgjm6zxasshp7k-pdfium-bin/lib']
_allow_system_search = True
_libpath = _find_library("pdfium", _libdirs, _allow_system_search)
_lib = ctypes.CDLL(_libpath)
```

With `--runtime-libdirs .` the first line is

```
_libdirs = ['.']
```

instead.

> supposing `${pdfium-bin}` is not available at runtime?

The above `_libdirs = ['/nix/store/q79y5h4af3vyi5kqbrwgjm6zxasshp7k-pdfium-bin/lib']` makes `${pdfium-bin}` available at runtime.

> Or if it is, you wouldn't have to copy in the binary below.

Right -- my understanding was that `--runtime-libdirs .` with the `cp` is for bundling the `.so` inside `pypdfium2`, while `--compile-libdirs "${pdfium-bin}/lib"` would be to reference it.

Nixpkgs would likely rather reference it (better build deduplication / sharing of dependencies), but I found that removing the `cp` has the inconvenient side effect that it produces this error:

```
Version uncertain: git repo not available.
Falling back to autorelease record.
Traceback (most recent call last):
  File "/build/source/nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 154, in <module>
    main()
  File "setup.py", line 150, in main
    run_setup(modnames, pl_name, pdfium_ver)
  File "setup.py", line 134, in run_setup
    assert_exists(ModuleDir_Raw, kwargs["package_data"]["pypdfium2_raw"])
  File "setup.py", line 76, in assert_exists
    assert False, f"Missing data files: {missing}"
AssertionError: Missing data files: ['libpdfium.so']
```

This is because of:

```python
    elif pl_name == ExtPlats.system:
        kwargs["package_data"]["pypdfium2_raw"] = [VersionFN, BindingsFN]
    else:
        # ...
        kwargs["package_data"]["pypdfium2_raw"] = [VersionFN, BindingsFN, libname]
```

I'm currently setting `PDFIUM_PLATFORM='prepared!linux_x64:6110`, not `PDFIUM_PLATFORM='prepared!system:6110`, so the lower branch is taken and it expects the `libname` (`.so`) in the `pypdfium2_raw` directory.

So far from the `pypdfium2` README I understood `system` to denotate "when the system has _built_ the `.so`", and `linux_x64` to be used when using the `pdfium` binary releases, but maybe "placing the binary relase `.so` in a system directory" also means I should better use `system` instead?

mara004 replied:
Yes, `system` is the correct one here. The latter meaning "in a system directory" is right. Explicit platforms like `linux_x64` imply bundling into pypdfium2's source tree, as the assertion hints. With `prepared!`, the target value is basically only used for the file inclusion mask.

Since I'm not used to the recipe format, I just wasn't sure about the scope of `${pdfium-bin}` (temporary to setup, or part of the install), and therefore not sure what you were trying to achieve.
But this clarifies `${pdfium-bin}` is part of the install, and you're wanting to load that at runtime, so no need to bundle an unused binary.

`--compile-libdirs` is used by ctypesgen to load the library at setup/bindings creation time to verify symbol presence, and does not impact install time library loading, which is controlled by `--runtime-libdirs` instead.
I believe you can also use `--universal-libdirs` if both match, but I haven't actually tested this yet.

And confusing though it may sound, you might want to add `--no-system-libsearch` to disable the `ctypes.util.find_library()` fallback, since you are passing the runtime libdir explicitly.
(Maybe the flag should be renamed to `--no-fallback-libsearch` to better reflect its meaning.)

Or if the library were contained in a "standard system location" as recognized by ctypes (e.g. `/usr/lib`), you could alternatively omit the explicit `--runtime-libdirs` and rely on ctypes to find it, but as NixOS library locations look quite different than on other distros, I don't think that would work, unless NixOS spoofs a different system layout (similar to conda), sets `$LD_LIBRARY_PATH` or something.

Note, if the binary is contained in a caller-given runtime libdir, pypdfium2 will of course depend on the continued presence of that dir, and load the library from there.
So if pdfium is updated and the directory's hash changes, you also have to rebuild pypdfium2 to link against the newer version of pdfium. Actually this is good because it enforces ABI safety.

mara004 commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,152 @@
+
+    cp "${pdfium-bin}/lib/libpdfium.so" src/pypdfium2_raw/libpdfium.so
+
+    export PDFIUM_PLATFORM=prepared!${pdfium.platform}:${toString pdfiumVersion_build}
```
In a bash shell I think you have to wrap this value in `'` because of the `!`.

nh2 replied:
It does work in this context, because `!` history expansion is off in non-interactive shells.

But it should definitely be quoted so people can easily copy-paste it in normal shells; will fix.

mara004 replied:
Good to know, thanks.

mara004 commented:
Thanks, pypdfium2 part basically looks good, except for the two spots noted above, which I don't think would work as-is.

nh2 commented:
I've switched to `system` and also added a doctr OCR test for PDFs that exercises the `pypdfium2` code path.

nh2 commented:
I force-pushed, also renaming `pytorch` -> `torch` so that hopefully the `ofborg-eval` CI passes.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2910

nh2 commented:
> Sorry BTW for acting a bit too rashly in our previous discussion. Anyway, I believe we've reached a good result now.

@mara004 No worries. Yes, I think the result is good, thanks for the offline packaging instructions!

mara004 commented:
Also note that you need to distribute the `LICENSE` file of the pdfium-binaries tarball, which contains the licenses for third-party components bundled into pdfium. I'm not sure if this is done already - again I don't understand the recipe format well enough to tell.

mara004 commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+      "patch": ${toString pdfiumVersion_patch},
+      "n_commits": 0,
+      "hash": null,
+      "origin": "pdfium-binaries/nixos",
```
It's quite unclear whether this should be `pdfium-binaries/nixos` or `system/nixos`, but that's our upstream fault. In theory `pdfium-binaries` implies bundling, but `system` doesn't include the meaning that the binary comes from pdfium-binaries.
Since a hashed, explicit runtime libdir is as good as bundling, I'd leave it as-is for now.
I plan to revise the version file (and how our code is writing it) to cleanly include all these meanings: library location (bundled/external), package flavor (pypa / conda / linux distro), and binary builder (pdfium-binaries / distro / other)

pbsds commented on pkgs/development/python-modules/doctr/default.nix:
```diff
@@ -0,0 +1,140 @@
+
+    # Torch dependencies from:
+    #     https://github.com/mindee/doctr/blob/v0.7.0/pyproject.toml#L62-L66
+    torch
+    torchvision
+    onnx
```
Would it make sense to move these deps to `passthru.optional-dependencies.torch`?

pbsds commented on pkgs/development/python-modules/doctr/default.nix:
```diff
@@ -0,0 +1,140 @@
+    homepage = "https://mindee.github.io/doctr/";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ chpatrick nh2 ];
+    platforms = [ "x86_64-linux" ]; # because `pypdfium2` is only packed for this so far
```
```suggestion
    inherit (pypdfium2.meta) platforms;
```

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+
+  # Contains more than a single dir; `builtins.fetchTarball` and `fetchzip` cannot handle that yet,
+  # so we need to manually unpack it below.
+  pdfiumPrebuiltTar = (fetchurl {
+    url = "https://github.com/bblanchon/pdfium-binaries/releases/download/chromium/${toString pdfiumVersion_build}/pdfium-${pdfium.platformFileInfix}.tgz";
+    hash = pdfium.hash;
+  });
+
+  pdfium-bin = runCommand "pdfium-bin" {} ''
+    mkdir "$out"
+    tar xf "${pdfiumPrebuiltTar}" --directory "$out"
+  '';
```
```suggestion
  pdfiumPrebuiltTar = (fetchzip {
    url = "https://github.com/bblanchon/pdfium-binaries/releases/download/chromium/${toString pdfiumVersion_build}/pdfium-${pdfium.platformFileInfix}.tgz";
    hash = pdfium.hash;
    stripRoot = false;
  };
```

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+    }:
+      buildPythonApplication rec {
+        pname = "ctypesgen";
+        version = "3a811c28160b7c742f97ee50bfe55b18d4021ca0";
```
```suggestion
        version = "1.1.1-unstable-2023-11-05";
```

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+          hash = "sha256-VqZ7Xn/7mrBjzIA7RUQt4WkrOttJMb4XmFWPpGtPfHY=";
+        };
+
+        # This package uses setuptools-scm to derive the version from the git repo (which we don't have),
+        # so use this environment variable to set it manually instead.
+        SETUPTOOLS_SCM_PRETEND_VERSION = "1.1.1+g${version}"; # we use the most recent tag and append the git version
```
```suggestion
        env.SETUPTOOLS_SCM_PRETEND_VERSION = "1.1.1+g${src.rev}"; # the most recent tag + git version
```

pbsds replied:
Even better if it could be also be based on `version`

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+        src = fetchFromGitHub {
+          owner = "pypdfium2-team"; # fork needed for `pypdfium2`
+          repo = "ctypesgen";
+          rev = version;
```
```suggestion
          rev = "3a811c28160b7c742f97ee50bfe55b18d4021ca0";
```

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+      assert lib.length jsonList == 4;
+      assert lib.all lib.isInt jsonList;
+      jsonList;
+  pdfiumVersion_major = lib.elemAt parsedNumericVersionComponents 0;
+  pdfiumVersion_minor = lib.elemAt parsedNumericVersionComponents 1;
+  pdfiumVersion_build = lib.elemAt parsedNumericVersionComponents 2;
+  pdfiumVersion_patch = lib.elemAt parsedNumericVersionComponents 3;
```
```suggestion
  pdfiumVersion.major = lib.elemAt parsedNumericVersionComponents 0;
  pdfiumVersion.minor = lib.elemAt parsedNumericVersionComponents 1;
  pdfiumVersion.build = lib.elemAt parsedNumericVersionComponents 2;
  pdfiumVersion.patch = lib.elemAt parsedNumericVersionComponents 3;
```

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+  '';
+
+  propagatedBuildInputs = [
+    build
+    setuptools-scm
+    numpy
+    pytest
+    pillow
```
```suggestion
    numpy
    pillow
```

Please don't propagate pytest, move it to `buildInputs`, also move build and setuptools-scm to `nativeBuildInputs`.

pbsds commented on pkgs/development/python-modules/pypdfium2/bin.nix:
```diff
@@ -0,0 +1,147 @@
+  meta = with lib; {
+    description = "Python bindings to PDFium";
+    homepage = "https://pypdfium2.readthedocs.io/";
+    license = with licenses; [ asl20 /* or */ bsd3 ];
```
Could you add a comment about the dual license?

pbsds reviewed (commented):
Result of `nixpkgs-review pr 265735` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>python310Packages.doctr</li>
    <li>python310Packages.doctr.dist</li>
    <li>python310Packages.pypdfium2-bin</li>
    <li>python310Packages.pypdfium2-bin.dist</li>
    <li>python311Packages.doctr</li>
    <li>python311Packages.doctr.dist</li>
    <li>python311Packages.pypdfium2-bin</li>
    <li>python311Packages.pypdfium2-bin.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.mplcursors</li>
    <li>python310Packages.mplcursors.dist</li>
    <li>python311Packages.mplcursors</li>
    <li>python311Packages.mplcursors.dist</li>
  </ul>
</details>


```
error: hash mismatch in fixed-output derivation '/nix/store/k84a5hkxr1cfabs922i5f8szkh553d4w-pdfium-linux-x64.tgz.drv':
         specified: sha256-d1tFlzJuh0q7gzwbVegPngxkiCdL/5kXZTOs/aXATlE=
            got:    sha256-aN7VSp+p7xuxcVn1uQuiIKqdV05El+WD3TWHmdq2gIU=
```

The build error might simply be a recompressed archive upstream? The suggestions below try to unpack the archive

nh2 commented:
`pypdfium2` has now also been packaged non-`bin` apparently in PR:

* #374461

That might simplify things here, not sure if we still should package `-bin` or not.

mara004 commented:
> `pypdfium2` has now also been packaged non-`bin` apparently in PR:
> 
> * [python3Packages.pypdfium2: init at 4.30.1 #374461](https://github.com/NixOS/nixpkgs/pull/374461)
> 
> That might simplify things here, not sure if we still should package `-bin` or not.

Decoupling the binaries is probably a good idea for distro packaging.
Unfortunately, it looks like they again tried to hack pypdfium2's `setupsrc`. I would prefer them to change to the caller-provided data files, see https://github.com/NixOS/nixpkgs/pull/374461/files#r2106065714
